lib/modules/watcher/index.js

```diff
@@ -0,0 +1,180 @@
+let chokidar = require('chokidar');
+let path = require('path');
+
+let fs = require('../../core/fs.js');
+
+// TODO: this should be receiving the config object not re-reading the
+// embark.json file
+class Watcher {
+  constructor(embark) {
+    this.logger = embark.logger;
+    this.events = embark.events;
+    this.fileWatchers = [];
+
+    this.events.setCommandHandler('watcher:start', () => this.start());
+    this.events.setCommandHandler('watcher:stop', () => this.stop());
+    this.events.setCommandHandler('watcher:restart', () => this.restart());
+  }
+
+  start() {
+    let self = this;
+    // TODO: should come from the config object instead of reading the file
+    // directly
+    let embarkConfig = fs.readJSONSync("embark.json");
+
+    this.watchAssets(embarkConfig, function () {
+      self.logger.trace('ready to watch asset changes');
+    });
+
+    this.watchContracts(embarkConfig, function () {
+      self.logger.trace('ready to watch contract changes');
+    });
+
+    this.watchContractConfig(embarkConfig, function () {
+      self.logger.trace('ready to watch contract config changes');
+    });
+
+    this.watchWebserverConfig(embarkConfig, function () {
+      self.logger.trace('ready to watch webserver config changes');
+    });
+
+    this.logger.info(__("ready to watch file changes"));
+  }
+
+  restart() {
+    this.stop();
+    this.start();
+  }
+
+  stop() {
+    this.fileWatchers.forEach(fileWatcher => {
+      if (fileWatcher.shouldClose) return;
+      if (fileWatcher.isReady) fileWatcher.close();
+      fileWatcher.shouldClose = true;
+    });
+  }
+
+  watchAssets(embarkConfig, callback) {
+    let self = this;
+    let appConfig = embarkConfig.app;
+    let filesToWatch = [];
+
+    for (let targetFile in appConfig) {
+      let files = appConfig[targetFile];
+      let fileGlob = files;
+
+      // workaround for imports issue
+      // so embark reacts to changes made in imported js files
+      // chokidar glob patterns only work with front-slashes
+      if (!Array.isArray(files)) {
+        fileGlob = path.join(path.dirname(files), '**', '*.*').replace(/\\/g, '/');
+      } else if (files.length === 1) {
+        fileGlob = path.join(path.dirname(files[0]), '**', '*.*').replace(/\\/g, '/');
+      }
+
+      filesToWatch.push(fileGlob);
+    }
+    filesToWatch = Array.from(new Set(filesToWatch));
+
+    this.watchFiles(
+      filesToWatch,
+      function (eventName, path) {
```

Comment:
You have used arrow function everywhere except here.
Out of curiosity, any reason for that?

Reply:
No particular reason. I honestly didn't make any code style changes in this module, but I'm happy to change that!

Reply:
Could we add a watcher for a DApp's `webpack.config.js` and `babel-loader-overrides.js` if they exist (no need to watch the default ones inside embark)? Also for the pipeline config?

Reply:
Yes, good point. I can extend this.

Reply:
Okay so I've added a watcher for the `webpack.config.js` inside the dapp, as well as a possible `config/pipeline.js` file. I'm not sure about the `babel-loader-overrides.js` though... Is that something that we expect people to create? Or is that something that gets ejected as well when we do an `embark eject-webpack`?

Reply:
It's something that gets ejected by `embark eject-webpack`.